File: java/sdk/src/test/java/org/acc/service/sdk/client/SettlementClientForShop.java
package org.acc.service.sdk.client;

import org.acc.service.sdk.data.NetWorkType;
import org.acc.service.sdk.utils.CommonUtils;
import org.json.JSONObject;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.nio.charset.StandardCharsets;

public class SettlementClientForShop extends SettlementClient {
    protected String shopId;

    public SettlementClientForShop(NetWorkType network, String privateKey, String shopId) {
        super(network, privateKey, shopId);

        this.shopId = shopId;
    }

    public String getSettlementManager() throws Exception {
        URI uri = new URI(relayEndpoint + "/v1/shop/settlement/manager/get/" + getShopId());
        HttpURLConnection conn = getHttpURLConnection(uri, "GET");
        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("managerId");
    }

    public String setSettlementManager(String managerId) throws Exception {
        byte[] message = CommonUtils.getSetSettlementManagerMessage(
                getShopId(),
                managerId,
                this.getShopNonceOf(getAddress()),
                this.getChainId()
        );
        String signature = CommonUtils.signMessage(this.credentials.getEcKeyPair(), message);

        URI uri = new URI(String.format("%s/v1/shop/settlement/manager/set", relayEndpoint));
        HttpURLConnection conn = getHttpURLConnection(uri, "POST");

        JSONObject body = new JSONObject();
        body.put("shopId", getShopId());
        body.put("account", getAddress());
        body.put("managerId", managerId);
        body.put("signature", signature);

        try (OutputStream output = conn.getOutputStream()) {
            output.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("txHash");
    }

    public String removeSettlementManager() throws Exception {
        byte[] message = CommonUtils.getRemoveSettlementManagerMessage(
                getShopId(),
                this.getShopNonceOf(getAddress()),
                this.getChainId()
        );
        String signature = CommonUtils.signMessage(this.credentials.getEcKeyPair(), message);

        URI uri = new URI(String.format("%s/v1/shop/settlement/manager/remove", relayEndpoint));
        HttpURLConnection conn = getHttpURLConnection(uri, "POST");

        JSONObject body = new JSONObject();
        body.put("shopId", getShopId());
        body.put("account", getAddress());
        body.put("signature", signature);

        try (OutputStream output = conn.getOutputStream()) {
            output.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("txHash");
    }

    public String getAgentOfRefund() throws Exception {
        URI uri = new URI(relayEndpoint + "/v1/agent/refund/" + getShopId());
        HttpURLConnection conn = getHttpURLConnection(uri, "GET");
        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("agent");
    }

    public String setAgentOfRefund(String agent) throws Exception {
        byte[] message = CommonUtils.getRegisterAgentMessage(
                getAddress(),
                agent,
                this.getLedgerNonceOf(getAddress()),
                this.getChainId()
        );
        String signature = CommonUtils.signMessage(this.credentials.getEcKeyPair(), message);

        URI uri = new URI(String.format("%s/v1/agent/refund", relayEndpoint));
        HttpURLConnection conn = getHttpURLConnection(uri, "POST");

        JSONObject body = new JSONObject();
        body.put("account", getAddress());
        body.put("agent", agent);
        body.put("signature", signature);

        try (OutputStream output = conn.getOutputStream()) {
            output.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("txHash");
    }

    public String getAgentOfWithdrawal() throws Exception {
        URI uri = new URI(relayEndpoint + "/v1/agent/withdrawal/" + getShopId());
        HttpURLConnection conn = getHttpURLConnection(uri, "GET");
        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("agent");
    }

    public String setAgentOfWithdrawal(String agent) throws Exception {
        byte[] message = CommonUtils.getRegisterAgentMessage(
                getAddress(),
                agent,
                this.getLedgerNonceOf(getAddress()),
                this.getChainId()
        );
        String signature = CommonUtils.signMessage(this.credentials.getEcKeyPair(), message);

        URI uri = new URI(String.format("%s/v1/agent/withdrawal", relayEndpoint));
        HttpURLConnection conn = getHttpURLConnection(uri, "POST");

        JSONObject body = new JSONObject();
        body.put("account", getAddress());
        body.put("agent", agent);
        body.put("signature", signature);

        try (OutputStream output = conn.getOutputStream()) {
            output.write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        JSONObject data = getJSONObjectResponse(conn);
        return data.getString("txHash");
    }

}

File: typescript/tests/helper/SettlementClientForShop.ts
import { CommonUtils } from "../../src";
import { SettlementClient } from "../../src/client/SettlementClient";
import { HTTPClient } from "../../src/network/HTTPClient";

import URI from "urijs";

/**
 *
 */
export class SettlementClientForShop extends SettlementClient {
    public async getSettlementManager(): Promise<string> {
        const agent = new HTTPClient({});
        const response = await agent.get(
            URI(this.endpoints.relay).directory("/v1/shop/settlement/manager/get").filename(this.shopId).toString()
        );
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }

        return response.data.data.managerId;
    }

    public async setSettlementManager(managerId: string): Promise<string> {
        const agent = new HTTPClient({});
        const nonce = await this.getShopNonceOf(this.wallet.address);
        const message = CommonUtils.getSetSettlementManagerMessage(
            this.shopId,
            managerId,
            nonce,
            await this.getChainId()
        );
        const signature = await CommonUtils.signMessage(this.wallet, message);
        const response = await agent.post(
            URI(this.endpoints.relay).directory("/v1/shop/settlement/manager/").filename("set").toString(),
            {
                shopId: this.shopId,
                account: this.wallet.address,
                managerId,
                signature,
            }
        );
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }

        return response.data.data.txHash;
    }

    public async removeSettlementManager(): Promise<string> {
        const agent = new HTTPClient({});
        const nonce = await this.getShopNonceOf(this.wallet.address);
        const message = CommonUtils.getRemoveSettlementManagerMessage(this.shopId, nonce, await this.getChainId());
        const signature = await CommonUtils.signMessage(this.wallet, message);
        const response = await agent.post(
            URI(this.endpoints.relay).directory("/v1/shop/settlement/manager/").filename("remove").toString(),
            {
                shopId: this.shopId,
                account: this.wallet.address,
                signature,
            }
        );
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }

        return response.data.data.txHash;
    }

    public async getAgentOfRefund(account?: string): Promise<string> {
        if (account === undefined) account = this.wallet.address;
        const client = new HTTPClient({});
        const response = await client.get(
            URI(this.endpoints.relay).directory("/v1/agent/refund/").filename(account).toString()
        );
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }

        return response.data.data.agent;
    }

    public async setAgentOfRefund(agent: string): Promise<string> {
        const client = new HTTPClient({});
        const nonce = await this.getLedgerNonceOf(this.wallet.address);
        const message = CommonUtils.getRegisterAgentMessage(this.wallet.address, agent, nonce, await this.getChainId());
        const signature = await CommonUtils.signMessage(this.wallet, message);
        const response = await client.post(URI(this.endpoints.relay).directory("/v1/agent/refund").toString(), {
            account: this.wallet.address,
            agent,
            signature,
        });
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }
        return response.data.data.txHash;
    }

    public async getAgentOfWithdrawal(account?: string): Promise<string> {
        if (account === undefined) account = this.wallet.address;
        const client = new HTTPClient({});
        const response = await client.get(
            URI(this.endpoints.relay).directory("/v1/agent/withdrawal/").filename(account).toString()
        );
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }
        return response.data.data.agent;
    }

    public async setAgentOfWithdrawal(agent: string): Promise<string> {
        const client = new HTTPClient({});
        const nonce = await this.getLedgerNonceOf(this.wallet.address);
        const message = CommonUtils.getRegisterAgentMessage(this.wallet.address, agent, nonce, await this.getChainId());
        const signature = await CommonUtils.signMessage(this.wallet, message);
        const response = await client.post(URI(this.endpoints.relay).directory("/v1/agent/withdrawal").toString(), {
            account: this.wallet.address,
            agent,
            signature,
        });
        if (response.data.code !== 0) {
            throw new Error(response.data.error?.message);
        }
        return response.data.data.txHash;
    }
}
